### ISSUE TITLE
[no-unused-expressions] extend to cover some functions

### ISSUE THREAD

**Issue by PSeitz** [typescript-eslint/typescript-eslint#2509]
There are some function calls which, where the return value should not be ignored (like some, indexOf, concat, map, etc.) or the function call doesn't make sense. Would be nice to cover this with the no-unused-expressions lint. I did test the snippets below on current master, they both don't fail.

```typescript
      function foo(arr) {
        if (arr) {
          arr.some(el == 'myval'); // should be: return arr.some(el == 'myval');
        }
        return undefined;
      }
```

```typescript
      function foo(arr) {
        var merged = arr.reduce(function (a, b) {
          a.concat(b);  // should be: return a.concat(b);
        });
      }
```

**Comment by bradzacher** [comment]
I have seen people do weird things like use `some` as a short-circuiting iterator (yes it's weird, yes I hate it, but people do do it).

I think that this would be better as a separate rule that's more generalised (for example #1314)

**Comment by PSeitz** [comment]
Now you mention it, I also saw `some` used like this. I think it's bad style, because it's hard to understand but could be optionally accepted.
The main difference between #1314 is, that I would like to focus on builtin methods where ignoring the value is very likely a bug. 

So I could imaging, that on #1314 we can control that normal return values functions are ignored, but that predefined list of function names are errors (which again could be extended or shortened, e.g. to cover `some`). What do you think?

**Comment by bradzacher** [comment]
Configuring a rule like this is rather difficult, as it is with any type-aware lint rule.
Names are entirely contextual. So you have to design a system for specifying the type name + method name pairs such that you can uniquely resolve them to a single type (i.e. you don't want to false-positive on a type just because it has the same name).

This is why I usually don't allow simple ignore lists in type-aware lint rules.

----

As an aside, my only problem with rules like this (which is why I've never bothered to implement them myself) is that they are going to be _super_ expensive to run.

TS types are lazily computed, so the more that a rule inspects, the slower that the rule runs. We already see this in rules like `no-floating-promises` - which inspects the type of every `ExpressionStatement`.

It's a high price to pay for a rarer bug vector.

**Comment by PSeitz** [comment]
Yes, since this is type-aware it should not be mixed with #1314, but be a seperate rule.

How should expensive functions handled? Flagged as expensive or moved to a seperate plugin?
I think it would make sense as part of a more extensive codescan, maybe on ci.

**Comment by bradzacher** [comment]
We can flag it in the readme for the rule for sure.

**Comment by jguddas** [comment]
I like this 👍 

If you think about it, there are a crazy amount of ways to have floating expressions.

```ts
function iContainFloatingEpressions() {
    'foo';
    550;
    [].slice(0);
    [].splice(0);
    (new Date()).getHours();
    (() => { console.log() });
}
```

**Comment by Josh-Cena** [comment]
This should be part of https://typescript-eslint.io/rules/no-unused-expressions/, no? We are essentially special-casing a subset of call expressions to be flagged.

**Comment by bradzacher** [comment]
It could be - it's just that it's quite a deviation from the rule.
It's a reasonably rare case that people leave behind such dead code - so I'm not sure if people would really want this turned on by default?

There's space here to track all function calls and report if the return value is unused.
The problem is that there is no way to track/declare side-effects in TS. Really the rule wants to report on "all function calls that are pure and their return value is not used" - but without a system for tracking purity that sort of rule would be full of false-positives

**Comment by Josh-Cena** [comment]
Yes, I don't think there's a way to generically check function calls, but conceptually, this is using some prior knowledge of built-in methods to assume their purity. I'm not sure if putting it as an option makes the most sense—`{ ignoreBuiltInPureFunctions: false }` which can gradually expand in scope over time? A rule like `no-unused-array-method-call` certainly seems like an ad-hoc rule for a very specific subset of use cases.

**Comment by bradzacher** [comment]
I would say that this still belongs as a separate rule for two reasons:

first it would require types and thus would be a breaking change to introduce it into `no-unused-expressions`.

second I think we want to leave `no-unused-expressions` open in case TS adds new syntax in future that could be an unused expression. For example #7748 - the rule should flag `Foo<T>`.
If we were to introduce type-aware behaviour into the extension rule then we would restrict people from having the basic checks for TS-specific syntax like this.

**Comment by bradzacher** [comment]
Coming back to this with an interesting musing.
Rust has a directive that can be annotated on all manner of things and it causes rust to do compile-time checks on that thing to ensure you don't have "unused expressions".

```rust
#[must_use]
struct MustUseThing;
fn returns_must_use_thing() -> MustUseThing {
    MustUseThing
}

#[must_use]
enum MustUseEnum {
    A,
}
fn returns_must_use_enum() -> MustUseEnum {
    MustUseEnum::A
}

#[must_use]
fn must_use_fn() -> &'static str {
    "hello"
}

fn main() {
    returns_must_use_thing(); // compiler error
    returns_must_use_enum(); // compiler error
    must_use_fn(); // compiler error
}
```
[rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=8c8d709e8cd4412ae5dbce91403a6cdc)

https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute

I wonder if we could create a rule doing something similar for TS -- for example build a rule around a new `@mustUse` JSDoc annotation (the impl would be very similar to `no-deprecated`).

**Comment by kirkwaiblinger** [comment]
Very related to https://github.com/typescript-eslint/typescript-eslint/issues/8255, but I guess there we're implicitly considering all disposables "mustUse"

**Comment by bradzacher** [comment]
The fun thing about this is that realistically it would be a relatively cheap rule to run on a codebas because you'd be able to focus solely on a small selector scope:
- `ExpressionStatement > :matches(CallExpression, NewExpression)`
- `ExpressionStatement > ChainExpression > :matches(CallExpression, NewExpression)`
- `ExpressionStatement > YieldExpression`
- `ExpressionStatement > AwaitExpression`
- You _probably_ would want completeness to check `void x()`, `x() as y`, `x() satisfies y` but that's small scope perf-wise

If you're using `no-floating-promises` then the types are already being computed so the rule would be "free".

**Comment by bradzacher** [comment]
Related TS issue: https://github.com/microsoft/TypeScript/issues/8240

**Comment by kirkwaiblinger** [comment]
Oh, a few things: 

1. I initially understood this as a no-floating-iterators rule, which would make sense for similar reasons to no-floating-disposables/no-floating-promises (resources not cleaned up). It being no-unused-array-predicates is a bit different, and just a special (but important!) case of the general question of when unused return values are an indication of a programmer error
2. As mentioned in this issue, whether unused array predicate returns are removable is dependent on whether the callback is pure, which we also have no control over. It *should* be, in a sane world, but people do all sorts of things. I'm ok with assuming the callback must be pure and that people writing side-effect-ful code should write it using `for...of`/`.foreach()`.
4. I'm skeptical that it's a good idea for us to try to define a general `@nodiscard`/`@mustuse` convention, for similar reasons to what's mentioned in the TS issue (https://github.com/microsoft/TypeScript/issues/8240#issuecomment-1960474970). *If* we tried to do such a thing, I think we'd want to let users configure an array of jsdoc tags to treat as nodiscard, and users might have their own internal convention in their codebase.
5. Arguably this is the type of thing that code review agents probably should be good at flagging.

Overall I'm probably -0.1.

Surely a non-type-aware version of the this rule (forbidding unused `.some()`/`.every()`/etc) would be almost exactly as effective as a type-aware version? Does one exist in the community?
